src/Algebra/Graph/Labelled/AdjacencyMap/Algorithm.hs

```diff
@@ -0,0 +1,51 @@
+module Algebra.Graph.Labelled.AdjacencyMap.Algorithm where
+
+import Algebra.Graph.Label
+import Data.Map.Strict (Map, (!))
+import Algebra.Graph.Labelled.AdjacencyMap
+
+import qualified Data.Set as Set
+import qualified Data.Map.Strict as Map
+
+-- TODO: Improve documentation for 'dijkstra'.
+-- | A generic Dijkstra algorithm that relaxes the list of edges
+-- based on the 'Dioid'.
+--
+-- If the 'Dioid' is 'Distance' (negative 'Dioid') the relaxation
```

Comment:
Are "negative" and "positive" dioids common terms? What is their precise definition?

Reply:
Couldn't you just use `ShortestDistance` and `WidestPath` from `Algebra.Graph.Label`?

https://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph-Label.html#t:ShortestPath
https://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph-Label.html#t:WidestPath

Reply:
> Are "negative" and "positive" dioids common terms? What is their precise definition?

I apologize, I actually was talking about the order of the underlying semi-ring.
If `1 > 0`, it would be a positive semi-ring. These definitions assume that there is a partial order defined on the set of elements which I believe I did not mention.

> Couldn't you just use ShortestDistance and WidestPath from Algebra.Graph.Label

This is not a generic algorithm but a Dijkstra algorithm. Dijkstra requires a priority heap depending on the order of the semi-ring. Even if one would use `ShortestDistance` the order of relaxation is decided by the heap. 
Using `ShortestDistance` with a max-heap or `WidestPath` with a min-heap would result in improper results. The type of heap is decided using the order of the semi-ring.

A generic single source shortest path algorithm does not require a priority heap but has a higher runtime complexity.

Reply:
@snowleopard I think one cannot define the order of the relaxation (min-heap vs max-heap) based on the type of Semiring. I believe we should take into account how `+` and `*` function as well.

Reply:
> These definitions assume that there is a partial order defined on the set of elements which I believe I did not mention.

Right. And I guess you are using `Ord` as this partial order?

Perhaps, you could use the following partial order defined for any idempotent semiring?

```
po x y = (x + y == y)
```

This is the order I use for algebraic graphs: `isSubgraphOf x y = (x + y == y)`.

> Please let me know if you know any good books/papers.

I think you already identified some very good papers on this topic. Not sure there is anything I can add to your list.